#5072 ca1305 ignore nullable invariant types

NewellClark opened the pull request:
Fix #5072

AndrewZu1337 commented:
@NewellClark  the your tests have been passed without fix. Looks like there isn't problem in this analyzer.

NewellClark commented:
> @NewellClark the your tests have been passed without fix. Looks like there isn't problem in this analyzer.

I did fix the issue brought up in #5072. I made two commits: one where I added failing tests, and another where I fixed the problem.

AndrewZu1337 commented:
> > @NewellClark the your tests have been passed without fix. Looks like there isn't problem in this analyzer.
> 
> @AndrewZu1337 I did fix the issue brought up in #5072. I made two commits: one where I added failing tests, and another where I fixed the problem.

I mean I ran the your failing tests without your fix and it's passed. So the analyzer handles `Nullable` versions of  types correctly without this fix.

NewellClark commented:
@AndrewZu1337 I ran the tests before implementing a fix, and they failed.

mavasani commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SpecifyIFormatProvider.cs:
```diff
@@ -86,12 +86,19 @@ protected override void InitializeWorker(CompilationStartAnalysisContext context
             var boolType = context.Compilation.GetSpecialType(SpecialType.System_Boolean);
             var guidType = context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemGuid);
+            var nullableTType = context.Compilation.GetSpecialType(SpecialType.System_Nullable_T);
+            INamedTypeSymbol? GetNullableType(INamedTypeSymbol? typeSymbol) => typeSymbol is not null ? nullableTType?.Construct(typeSymbol) : null;
```
Kindly move the local function towards the end of the method.

NewellClark replied:
I ended up removing the local function.

mavasani reviewed (approved):
Please retarget to `main` branch.

Youssef1313 commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/SpecifyIFormatProviderTests.cs:
```diff
@@ -945,6 +945,36 @@ End Class
 ");
         }
 
+        [Fact]
+        public async Task CA1305_NullableInvariantTypes_NoDiagnostic()
```
Can you add a workitem here?

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SpecifyIFormatProvider.cs:
```diff
@@ -86,12 +86,19 @@ protected override void InitializeWorker(CompilationStartAnalysisContext context
+            INamedTypeSymbol? GetNullableType(INamedTypeSymbol? typeSymbol) => typeSymbol is not null ? nullableTType?.Construct(typeSymbol) : null;
 
             var builder = ImmutableHashSet.CreateBuilder<INamedTypeSymbol>();
+
             builder.AddIfNotNull(charType);
+            builder.AddIfNotNull(GetNullableType(charType));
             builder.AddIfNotNull(boolType);
+            builder.AddIfNotNull(GetNullableType(boolType));
             builder.AddIfNotNull(stringType);
             builder.AddIfNotNull(guidType);
+            builder.AddIfNotNull(GetNullableType(guidType));
```
Would it be better to do the fix in `IsValidToStringCall`? (i.e, instead of constructing `char?`, `bool?`, and `Guid?`, just try to find `Nullable<SomeType>` and see if `SomeType` is in the hashset.

NewellClark replied:
Good idea. I went ahead and did that.
